Fix PostgreSQL `Cidr#change?` to compare with address prefix

taketo1113 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created because fixes #51582

### Detail

This Pull Request changes to override`ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr#changed?`.
It added to compare address prefix of IPAddr class.


### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

I’m proposing to fix `IPAddr#==` comparing address prefix for the cause of this problem.
https://github.com/ruby/ipaddr/pull/69

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

yahonda commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:
```diff
@@ -28,6 +28,11 @@ def serialize(value)
           end
 
+          def changed?(old_value, new_value, _new_value_before_type_cast)
+            old_value.class != new_value.class ||
```
I'm wondering if it needs `old_value.class != new_value.class` here. I assume it comes from https://github.com/rails/rails/commit/dbdd68b64b781ce3c3363ffbc223730138c49a33 , Cidr `cast_value` method looks like taking care of it https://github.com/rails/rails/blob/734719d8af033062cb7088a1353be6276659e3c1/activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb#L31-L43

taketo1113 replied:
@yahonda Sure I think too this is not necessary.
I have removed `old_value.class != new_value.class`.

taketo1113 committed:
Fix PostgreSQL Cidr#change? when changed address prefix. Fix #51582

Update activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb

Co-authored-by: Yasuo Honda <yasuo.honda@gmail.com>

Updated to keep the argument order

Added note to remove `changed?` method when `IPAddr#==` compares `IPAddr#prefix`

yahonda commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:
```diff
@@ -28,6 +28,10 @@ def serialize(value)
           end
 
+          def changed?(old_value, new_value, _new_value_before_type_cast)
+            new_value != old_value || new_value.prefix != old_value.prefix
```
```suggestion
            super || new_value.prefix != old_value.prefix
```

yahonda replied:
I think using `super` makes Cidr#changed? requirement clear. https://github.com/rails/rails/blob/734719d8af033062cb7088a1353be6276659e3c1/activemodel/lib/active_model/type/value.rb#L84-L86

yahonda replied:
I just prefer this to keep the argument order inside the method. old_value, then new_value.
```ruby
super || old_value.prefix != new_value.prefix
```

taketo1113 replied:
Thank you for your advices.
It has fixed to adopt `super` and keeping the argument order.

justinko commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/cidr.rb:
```diff
@@ -28,6 +28,10 @@ def serialize(value)
             end
           end
 
```
```suggestion
TODO: Remove when IPAddr#== compares IPAddr#prefix
```

taketo1113 replied:
Added todo comment to remove method and issues url.

yahonda commented:
Backported to the `7-1-stable` branch via https://github.com/rails/rails/commit/887210f8321d78ba9cdba9368a338c9411632f8b

vipulnsward commented:
@yahonda This needs a changelog as its a change in behavior?

yahonda commented:
Let me add a change log entry for this change. Thanks for the heads-up.

yahonda commented:
Opened https://github.com/rails/rails/pull/51742 to add a changelog entry for this change.

chaadow commented:
@yahonda @taketo1113 I'm on Rails 7-1 stable. and this makes my production app crash under certain circumstances. 
<img width="851" alt="image" src="https://github.com/rails/rails/assets/1371733/c14f58f3-8145-4915-895d-f499e2153805">

when `#save` is called on an activerecord model. here is the stack trace: 
<img width="1251" alt="image" src="https://github.com/rails/rails/assets/1371733/733563ff-e7b3-4876-8175-75499dbaa9ca">

chaadow commented:
( I do not have a CIDR column in any of my tables)

chaadow commented:
Honestly I wish I could reproduce the bug (it's in production) but it's so mystical. all I can say is, it's related to this PR and it was done on a "ActiveRecord#save" call. on an **unpersisted instance**

(probably the reason why `old_value.prefix` is nil ?)

but what's mystical to me is why it goes through the `changed?` call here even though I have 0 CIDR columns

chaadow commented:
I feel the suggestion in #51582 here would fix my case:
```ruby
class ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Cidr
  def changed?(old_value, new_value, _new_value_before_type_cast)
     old_value.nil? != new_value.nil? || (!old_value.nil? && !old_value.eql?(new_value))
  end
end
```

or at least the part that checks that it's not nil

chaadow commented:
@taketo1113 @yahonda I finally managed to recreate the issue

It's due to the fact **that `inet` inherits from CIDR**. 

this part helped me get less confused and better investigate and reproduce the issue

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  # If you want to test against edge Rails replace the previous line with this:
  gem "rails", github: "rails/rails", branch: "main"

  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

db_config = {
  adapter: 'postgresql',
  database: 'inet_cidr_test',
  host: 'localhost', # Change this to your PostgreSQL host
  port: 5432,
  username: 'postgres', # Change this to your PostgreSQL username
  password: 'postgres'  # Change this to your PostgreSQL password
}

ActiveRecord::Base.establish_connection(db_config.except(:database))
ActiveRecord::Base.connection.drop_database(db_config[:database]) rescue nil
ActiveRecord::Base.connection.create_database(db_config[:database])

ActiveRecord::Base.establish_connection(db_config)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.inet :ip_address
  end
end

class User < ActiveRecord::Base
end


class BugTest < Minitest::Test
  def test_ip_address_with_empty_string
    User.create!(ip_address: '')
  end
end
```

carlosantoniodasilva commented:
@chaadow if you haven't already, please open up a separate/new issue to help track this.

taketo1113 commented:
I confirmed the reproduce issue and created PR (#51758)
